File: Microsoft.Toolkit.Uwp.Input.GazeInteraction/GazeElement.cpp
//Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license.
//See LICENSE in the project root for license information.

#include "pch.h"
#include "GazeElement.h"
#include "DwellProgressEventArgs.h"
#include "StateChangedEventArgs.h"
#include "DwellInvokedRoutedEventArgs.h"
#include "winrt/Windows.Foundation.h"
#include "GazeElement.g.cpp"

namespace winrt::Microsoft::Toolkit::Uwp::Input::GazeInteraction::implementation
{
    winrt::event_token GazeElement::StateChanged(Windows::Foundation::EventHandler<Microsoft::Toolkit::Uwp::Input::GazeInteraction::StateChangedEventArgs> const& handler)
    {
        return m_stateChanged.add(handler);
    }
    void GazeElement::StateChanged(winrt::event_token const& token) noexcept
    {
        m_stateChanged.remove(token);
    }
    winrt::event_token GazeElement::DwellProgressFeedback(Windows::Foundation::EventHandler<Microsoft::Toolkit::Uwp::Input::GazeInteraction::DwellProgressEventArgs> const& handler)
    {
        return m_dwellProgressFeedback.add(handler);
    }
    void GazeElement::DwellProgressFeedback(winrt::event_token const& token) noexcept
    {
        m_dwellProgressFeedback.remove(token);
    }
    winrt::event_token GazeElement::Invoked(Windows::Foundation::EventHandler<Microsoft::Toolkit::Uwp::Input::GazeInteraction::DwellInvokedRoutedEventArgs> const& handler)
    {
        return m_invoked.add(handler);
    }
    void GazeElement::Invoked(winrt::event_token const& token) noexcept
    {
        m_invoked.remove(token);
    }

    void GazeElement::RaiseStateChanged(Windows::Foundation::IInspectable const& sender, Microsoft::Toolkit::Uwp::Input::GazeInteraction::StateChangedEventArgs const& args)
    {
        m_stateChanged(sender, args);
    }

    void GazeElement::RaiseInvoked(Windows::Foundation::IInspectable const& sender, Microsoft::Toolkit::Uwp::Input::GazeInteraction::DwellInvokedRoutedEventArgs const& args)
    {
        m_invoked(sender, args);
    }

    bool GazeElement::RaiseProgressFeedback(Windows::Foundation::IInspectable const& sender, Microsoft::Toolkit::Uwp::Input::GazeInteraction::DwellProgressState const& state, Windows::Foundation::TimeSpan const& elapsedTime, Windows::Foundation::TimeSpan const& triggerTime)
    {
        auto args = winrt::make<DwellProgressEventArgs>(state, elapsedTime, triggerTime);
        m_dwellProgressFeedback(sender, args);
        return args.Handled();
    }
}

File: CommunityToolkit.WinUI.Input.GazeInteraction/GazeElement.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using Microsoft.UI.Xaml;

namespace CommunityToolkit.WinUI.Input.GazeInteraction
{
    /// <summary>
    /// Surrogate object attached to controls allowing subscription to per-control gaze events.
    /// </summary>
    public class GazeElement : DependencyObject
    {
        /// <summary>
        /// This event is fired when the state of the user's gaze on a control has changed
        /// </summary>
        public event EventHandler<StateChangedEventArgs> StateChanged;

        /// <summary>
        /// This event is fired when the user completed dwelling on a control and the control is about to be invoked by default. This event is fired to give the application an opportunity to prevent default invocation
        /// </summary>
        public event EventHandler<DwellInvokedRoutedEventArgs> Invoked;

        /// <summary>
        /// This event is fired to inform the application of the progress towards dwell
        /// </summary>
        public event EventHandler<DwellProgressEventArgs> DwellProgressFeedback;

        internal void RaiseStateChanged(object sender, StateChangedEventArgs args)
        {
            StateChanged?.Invoke(sender, args);
        }

        internal void RaiseInvoked(object sender, DwellInvokedRoutedEventArgs args)
        {
            Invoked?.Invoke(sender, args);
        }

        internal bool RaiseProgressFeedback(object sender, DwellProgressState state, TimeSpan elapsedTime, TimeSpan triggerTime)
        {
            var args = new DwellProgressEventArgs(state, elapsedTime, triggerTime);
            DwellProgressFeedback?.Invoke(sender, args);
            return args.Handled;
        }
    }
}